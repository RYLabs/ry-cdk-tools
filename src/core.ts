export * from "./base_stacks/base_stack";
export { default as BaseStack } from "./base_stacks/base_stack";
export { default as Conventions } from "./utils/conventions";
