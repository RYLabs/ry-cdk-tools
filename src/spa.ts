export { default as SpaPipelineStack } from "./stacks/spa_pipeline_stack";
export * from "./stacks/spa_pipeline_stack";
