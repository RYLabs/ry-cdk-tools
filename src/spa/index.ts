export { default as S3SpaPipelineStack } from "../stacks/s3_spa_pipeline_stack";
export * from "../stacks/s3_spa_pipeline_stack";
export { default as S3Spa } from "../spa/s3_spa";
export * from "../spa/s3_spa";
export { default as CloudFrontConfig } from "./cloudfront_config";
export * from "./cloudfront_config";
