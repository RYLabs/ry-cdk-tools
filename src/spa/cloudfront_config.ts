import { Construct, CfnOutput } from '@aws-cdk/core';
import { IBucket } from '@aws-cdk/aws-s3';
import { IHostedZone, ARecord, RecordTarget } from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets/lib';
import {
  CloudFrontWebDistribution,
  SSLMethod,
  SecurityPolicyProtocol
} from '@aws-cdk/aws-cloudfront';

export interface CloudFrontConfigProps {
  subDomain?: string;
  /**
   * certificateArn: "arn:aws:acm:region:123456789012:certificate/12345678-1234-1234-1234-123456789012"
   * This value can be entered manually from your own certficate stored with ACM
   * or you can create Route53AcmSslStack.  
   */
  certificateArn: string;
  /**
   *  sourceBucket: IBucket
   *  The S3 bucket where your live project is located.
   */
  sourceBucket: IBucket;
  /**
   *  hostedZone: IHostedZone
   *  Route53 HostedZone created via Route53AcmSslStack
   */
  hostedZone: IHostedZone;
}

export default class CloudFrontConfig extends Construct {
  constructor(scope: Construct, id: string, props: CloudFrontConfigProps) {
    super(scope, id);

    const {
      subDomain = "www",
      certificateArn,
      sourceBucket,
      hostedZone,
    } = props;

    const fullDomain = subDomain + "." + hostedZone.zoneName;
    
    const distribution = new CloudFrontWebDistribution(this, `projectDistribution`, {
      aliasConfiguration: {
        acmCertRef: certificateArn,
        names: [fullDomain],
        sslMethod: SSLMethod.SNI,
        securityPolicy: SecurityPolicyProtocol.TLS_V1_2_2018
      },
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: sourceBucket,
          },
          behaviors: [ {isDefaultBehavior: true} ]
        },
      ],
    });

    new ARecord(this, 'SiteAliasRecord', {
      recordName: fullDomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
      zone: hostedZone,
    });

    new CfnOutput(this, 'Project HTTPS URL', {
      value: 'https://' + fullDomain,
      description: 'Project URL using CloudFront Route53 for SSL'
    });
  }
}
