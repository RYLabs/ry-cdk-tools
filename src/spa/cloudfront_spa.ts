import { Construct, RemovalPolicy } from "@aws-cdk/core";
import { DnsValidatedCertificate } from '@aws-cdk/aws-certificatemanager';
import { IHostedZone, ARecord, RecordTarget } from '@aws-cdk/aws-route53';
import { CloudFrontWebDistribution, SSLMethod, SecurityPolicyProtocol } from '@aws-cdk/aws-cloudfront';
import { Bucket } from "@aws-cdk/aws-s3";
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets/lib';

export interface CloudFrontSpaProps {
  // prefix for website address www. api. test.
  subDomain: string;
  hostedZone: IHostedZone;
}

// Single page application using Cloudfront
export default class CloudfrontSpa extends Construct {
  bucket: Bucket;
  websiteUrl: string;

  // Construct needed to be passed for the creation of resources.
  constructor(scope: Construct, id: string, props: CloudFrontSpaProps){
    super(scope, id);

    const {
      hostedZone,
      subDomain,
    } = props

    this.bucket = new Bucket(this, "siteBucket", {
      bucketName: `${id.toLowerCase()}-cloudfrontspa`,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const fullDomain = subDomain + "." + hostedZone.zoneName

    const cert = new DnsValidatedCertificate(scope, "httpsCert", {
      domainName: `*.${hostedZone.zoneName}`,
      hostedZone,

      // Cloudfront always looks for certs in us-east-1
      region: "us-east-1"
    });

    const distribution = new CloudFrontWebDistribution(scope, 'MyDistribution', {
      aliasConfiguration: {
        acmCertRef: cert.certificateArn,
        names: [fullDomain],
        sslMethod: SSLMethod.SNI,
        securityPolicy: SecurityPolicyProtocol.TLS_V1_2_2018
      },
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: this.bucket
          },
          behaviors : [ {isDefaultBehavior: true}]
        }
      ]
    });

    new ARecord(scope, 'SiteAliasRecord', {
      recordName: fullDomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
      zone: hostedZone,
    });

    this.websiteUrl = `https://${subDomain}.${hostedZone.zoneName}`;
  }
} 