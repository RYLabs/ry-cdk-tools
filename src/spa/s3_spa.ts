import { Construct, RemovalPolicy } from "@aws-cdk/core";
import { Bucket } from "@aws-cdk/aws-s3";

export interface S3SpaProps {
  bucketName: string;
}

// Single page application using S3
export default class S3Spa extends Construct {
  bucket: Bucket;

  get websiteUrl(): string {
    return this.bucket.bucketWebsiteUrl;
  }

  // Construct needed to be passed for the creation of resources.
  constructor(scope: Construct, id: string, props: S3SpaProps) {
    super(scope, id);

    const { bucketName } = props;

    this.bucket = new Bucket(this, "siteBucket", {
      bucketName,
      websiteIndexDocument: "index.html",

      // re-render the index doc to support pushState
      websiteErrorDocument: "index.html",

      publicReadAccess: true,
      removalPolicy: RemovalPolicy.DESTROY,
    });
  }
}
