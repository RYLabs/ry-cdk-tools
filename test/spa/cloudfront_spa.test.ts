import { SynthUtils } from "@aws-cdk/assert";
import { Stack } from "@aws-cdk/core";
import { HostedZone } from "@aws-cdk/aws-route53";
import { CloudfrontSpa } from "../../lib/spa";
import { Bucket } from "@aws-cdk/aws-s3";

test("creates a single page application", () => {
  const stack = new Stack();
  const hostedZone = new HostedZone(stack, "hostedZone", {
    zoneName: "example.com",
  });
  const bucket = new Bucket(stack, 'bucket', {
  });

  new CloudfrontSpa(stack, "spa", {
    subDomain: "test-site",
    certificateArn: "string",
    sourceBucket: bucket,
    domainName: "example.com",
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
