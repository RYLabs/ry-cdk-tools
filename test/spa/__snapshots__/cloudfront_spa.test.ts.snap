// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a single page application 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2ArtifactHash652C125C": Object {
      "Description": "Artifact hash for asset \\"19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2\\"",
      "Type": "String",
    },
    "AssetParameters19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2S3BucketFCCD3A76": Object {
      "Description": "S3 bucket for asset \\"19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2\\"",
      "Type": "String",
    },
    "AssetParameters19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2S3VersionKey07AF06B6": Object {
      "Description": "S3 key for asset version \\"19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyDistributionCFDistributionDE147309": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "test-site.example.com",
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "spasiteBucket0C2DE9A5",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {},
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": Object {
              "Fn::GetAtt": Array [
                "httpsCertCertificateRequestorResourceF9444076",
                "Arn",
              ],
            },
            "MinimumProtocolVersion": "TLSv1.2_2018",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "SiteAliasRecord3C0AF5BF": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "MyDistributionCFDistributionDE147309",
              "DomainName",
            ],
          },
          "HostedZoneId": "Z2FDTNDATAQYW2",
        },
        "HostedZoneId": Object {
          "Ref": "hostedZone861779BD",
        },
        "Name": "test-site.example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "hostedZone861779BD": Object {
      "Properties": Object {
        "Name": "example.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "httpsCertCertificateRequestorFunction9825D024": Object {
      "DependsOn": Array [
        "httpsCertCertificateRequestorFunctionServiceRoleDefaultPolicy434DCDC9",
        "httpsCertCertificateRequestorFunctionServiceRoleF6B1B640",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2S3BucketFCCD3A76",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2S3VersionKey07AF06B6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters19e461d2ff1a5b90438fed6ceee4c197d7efee8712a6f76d85b501ab20bfb1a2S3VersionKey07AF06B6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.certificateRequestHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "httpsCertCertificateRequestorFunctionServiceRoleF6B1B640",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "httpsCertCertificateRequestorFunctionServiceRoleDefaultPolicy434DCDC9": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:changeResourceRecordSets",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/",
                    Object {
                      "Ref": "hostedZone861779BD",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "httpsCertCertificateRequestorFunctionServiceRoleDefaultPolicy434DCDC9",
        "Roles": Array [
          Object {
            "Ref": "httpsCertCertificateRequestorFunctionServiceRoleF6B1B640",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "httpsCertCertificateRequestorFunctionServiceRoleF6B1B640": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "httpsCertCertificateRequestorResourceF9444076": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "*.example.com",
        "HostedZoneId": Object {
          "Ref": "hostedZone861779BD",
        },
        "Region": "us-east-1",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "httpsCertCertificateRequestorFunction9825D024",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "spasiteBucket0C2DE9A5": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": "spa-cloudfrontspa",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
